### ISSUE TITLE
Use of the transition-delay property appears to break swup

### ISSUE THREAD

**Issue by smedstadc** [gmrchk/swup#84]
Hi, I've been fiddling with swup to evaluate it for a project where it may be useful. So far I think it's pretty cool.

One thing I noticed is the use of the `transition-delay` property to stagger the entrance of elements appears to break page transitions. The page gets stuck in such a state that the user can't navigate and nothing appears.

I ran into this behavior in a bare bones project adapted from the examples in the README. The only additions were a second page to visit, a variant of the `.transition-fade` class with a short delay, and a static navbar that is always visible.

If the user clicks a different navbar link before the transition finishes, swup handles it just fine. If one of the transition classes on the page has a `transition-delay` property assigned to it, then you get the bad behavior I mentioned above.

Is this a limitation of swup, or a bug? 

If you need more information I can put together a clean example.

Thanks!

**Comment by gmrchk** [comment]
Hi @smedstadc, this sounds like a bug in a demo, as swup has no problem with waiting for transition end of delayed elements. Looks like you don't have a "hidden" state defined for at least one of the elements with class `transition-*`. This makes swup wait for the animation to finish, but that animation never even starts. In that case, swup doesn't throw any error (discussed [here](https://github.com/gmrchk/swup/issues/85#issuecomment-445802178)) and the only indication is the debug mode getting stuck at a certain point in the page transition lifecycle.

Of course, maybe I didn't understand the description correctly, and some code or live demo would be awesome if your problem persists.

**Comment by smedstadc** [comment]
Debug mode getting stuck at a point int he transition cycle sounds like what I was seeing. I'll double check my hidden state and see if that was the problem. I'll post an example if it was not.

**Comment by smedstadc** [comment]
After you mentioned the hidden state, I thought that I had forgotten to write a class for the new hidden state or there was a typo in one that I missed because I was looking for more complicated problems. Neither seems to be the case.

I think what happens is when a user starts a page transition, then clicks a different link after the immediate animation starts, but before the delayed animation starts, the cycle can get stuck. This is easier to reproduce with a longer delay.

You can observe the behavior in this project: https://github.com/smedstadc/swup-toy

There still may be something I've missed. I'm just not sure.

**Comment by gmrchk** [comment]
Thank you for the demo, it's really helpful. 

I see what's the problem now. The problem occurs if the second click happens after swup replaces the content, but before the delayed transition starts. This happens because swup will start another transition (which is an intended behavior), and animate the page *out* again, but the delayed transition hasn't started yet, and never will, as it is already in the *out* state.

I'm not sure swup should consider such a scenario, as it would always have to check whether the transition started, or if it ever will, and only after it starts it would do its thing and wait for the transition to end. Primarily because I'm not aware of some native way to do that, but also because it would add on a complexity of the code.

But there are some solutions to this, even if you want to use the delayed transitions:

1. Define your CSS as you did, but don't use a `transition-*` class for the delayed transition. This means that swup doesn't consider those elements at all and doesn't wait for them to finish the animation, but the elements still get animated based on `is-animating` class. If the animations have at least approximately the same duration (with the delay), this should not be an issue.    

    In fact, I would recommend to only have one element with `transition-*` class, and just use its timing to set the animation duration. The fact that swup will wait for any number of elements doesn't mean we should abuse the feature. Any other element can still be animated based on `is-animating` class and have the animation the same duration or shorter than the main transition element. This makes changing the duration super easy, ensures that all pages have the same duration for the transition and forces you to set some reasonable timing (you probably don't want to make a user wait for 5 seconds to show him the next page) and stick with it.

     [Here](https://plzen.cz.panasonic.com/) you can see an example of a page with only one `transition-fade` element. There is a bunch of elements still animated using `is-animating` class, but not being waited for by swup. Some elements even have a delayed transition, like the titles and the texts underneath.

2. Block any clicks on links while the page is in transition. This should be as simple as this and should work just fine as well.
    ```css
    html.is-changing a {
        pointer-events: none;
    }
    ```


3. Don't base the timing on CSS at all and use [swupjs](https://github.com/gmrchk/swupjs) instead.

**Comment by smedstadc** [comment]
Thanks for the great answer. I don't spend as much time on the front-end side of things, so the extra tips are helpful.

I think `1.` is the best option, because it avoids the extra time at the beginning of each transition. I walked away from my computer for the day after I posted the example repository in this thread. After I wasn't staring at a screen anymore I realized I could just drop the `transition-*` from the troublesome class and swup wouldn't watch it anymore. It was a real forehead-smack moment.

I don't normally think of the `pointer-events:` property, but I'm glad `2.` is an option if I want to stubbornly insist on waiting for the delays.
